lib/path/default.nix

```diff
@@ -94,6 +95,47 @@ let
 
 in /* No rec! Add dependencies on this file at the top. */ {
 
+  /* Append a subpath string to an absolute path value, the result is an absolute path value.
+
+    Like `path + ("/" + string)` but safer, because it errors for potentially unexpected results.
+    More specifically, it checks that the first argument is a path value,
+    and that the second argument is a valid subpath string, see `lib.path.subpath.isValid`.
+
+    Type:
+      append :: Path -> String -> Path
+
+    Examples:
+      append /foo "bar/baz"
+      => /foo/bar/baz
+
+      # subpaths don't need to be normalised
+      append /foo "./bar//baz/./"
+      => /foo/bar/baz
+
+      # can append to root directory
+      append /. "foo/bar"
+      => /foo/bar
+
+      # first argument needs to be a Nix path value
+      append "/foo" "bar"
+      => <error>
+
+      # second argument needs to be a valid subpath string
+      append /foo /bar
+      => <error>
+      append /foo ""
+      => <error>
+      append /foo "/bar"
+      => <error>
+      append /foo "../bar"
+      => <error>
+  */
+  append = base: subpath:
+    assert assertMsg (isPath base)
+      "lib.path.append: The first argument is of type ${builtins.typeOf base}, but a path value was expected";
+    assert assertMsg (subpathInvalidReason subpath == null)
+      "lib.path.append: Second argument is not a valid subpath string: ${subpathInvalidReason subpath}";
```

Comment:
```suggestion
    assert assertMsg (subpathInvalidReason subpath == null) ''
      lib.path.append: Second argument is not a valid subpath string:
          ${subpathInvalidReason subpath}
    '';
```

Reply:
Previously discussed in https://github.com/NixOS/nixpkgs/pull/205190#discussion_r1044936046. At least `path.normalise` and all the other library functions do not use newlines in throws, I think we should stay consistent with that.

Reply:
Exceptions are only ever presented one at a time, so the need for consistency isn't really there.
Nix itself uses multi-line messages, so feel free to do the same.

Reply:
In terms of consistency, we can of course fix up the code in multiple places in a follow-up. I'd just prefer to start while creating new code instead of piling up more work for later, which is more probable to get glanced over.

Reply:
Fair enough, I'm using newlines now